### PR TITLE
♻️ Remove incremental

### LINKED ISSUE #308: Remove dependency on incremental in favor of general compatible solution
https://github.com/twisted/towncrier/pull/297#pullrequestreview-556014893
> For `get_version` I think that we should have this logic
> 
> See if version is a tuple and join them with a dot.
> For anything else, just call `str(version_object)` and strip it.
> 
> In this way it will work with incremental and any other version objects.
> 
> towncrier shouldn't care about comparing versions...
> 
> It can try to be smart and detect that you are not adding a version older than the latest one in the file... but I think that is too much...and again it should be implemented with a naive compare
> 
> So I am +1 for remove the incremental dependency :) and use simple version strings for towncrier :)
> 
> I am using a fork of towncrier in which incremental is removed as at least 3 years ago, incremental was not playing nice with private PYPI repos due to a bug in setup tools

Sounds like we could remove the dependency specifically on incremental.  Give it a shot if we get a chance.

### PR DESCRIPTION
**Pull request by sigma67**
# Description

Fixes #308

This removes the dependency on `incremental` on `towncrier` itself, while keeping support for detecting `incremental` version in projects for which towncrier is used to build the release notes.

# Checklist

* [ ] Make sure changes are covered by existing or new tests.
* [ ] For at least one Python version, make sure local test run is green.
* [ ] Create a file in `src/towncrier/newsfragments/`. Describe your
  change and include important information. Your change will be included in the public release notes.
* [ ] Make sure all GitHub Actions checks are green (they are automatically checking all of the above).
* [ ] Ensure `docs/tutorial.rst` is still up-to-date.
* [ ] If you add new **CLI arguments** (or change the meaning of existing ones), make sure `docs/cli.rst` reflects those changes.
* [ ] If you add new **configuration options** (or change the meaning of existing ones), make sure `docs/configuration.rst` reflects those changes.

### CODE REVIEW

File: RELEASE.rst
Comment (adiroiban) [inline RELEASE.rst]: I was hoping that we can set the version in single place, like `src/towncrier/_version.py` and the build system can automatically extract the version from there.

File: docs/conf.py
Comment (adiroiban) [inline docs/conf.py]: Why not just use `from towncrier import __verison__` or something similar ?

File: src/towncrier/_project.py
Comment (adiroiban) [inline src/towncrier/_project.py]: I this change needed ?

I was thinking that this was sorted out in https://github.com/twisted/towncrier/pull/502

File: src/towncrier/_shell.py
Comment (adiroiban) [inline src/towncrier/_shell.py]: Why not keep the import ?

It's just that now `__version__` will be a simple text.

**Review by adiroiban — COMMENTED** [review]
Many thanks for picking up on this.

I just did a quick review before approving the CI run.

I have not yet fully reviewed the automated tests.

File: docs/conf.py
Comment (sigma67) [reply docs/conf.py]: The changes in `doc` were recycled but now they are no longer relevant. Reverted

File: RELEASE.rst
Comment (sigma67) [reply RELEASE.rst]: The changes in `doc` were recycled but now they are no longer relevant. Reverted

File: src/towncrier/_project.py
Comment (sigma67) [reply src/towncrier/_project.py]: Right, I think it's just a merge artifact. Reverted

File: src/towncrier/_shell.py
Comment (sigma67) [reply src/towncrier/_shell.py]: done

File: src/towncrier/test/test_packaging.py
Comment (adiroiban) [inline src/towncrier/test/test_packaging.py:10]: I am not sure why this change is needed.

The previous test look good enough to me. :)

I am misssing something ?

File: src/towncrier/test/test_project.py
Comment (adiroiban) [inline src/towncrier/test/test_project.py]: Maybe something like this ?


```suggestion
        An excpettion is raised when the version could not be detected.
        For this test we use an incremental-like object,
        that has the `base` method but that method
        does not matches the return type for `incremental`.
```

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks for the changes.
They look good.

I left a few minor comments.

In order to merge this, the `mypy` checks must pass.

Also the coverage must pass

```
src/towncrier/_project.py      52      1     18      1    97%   90, 104->exit
```

File: src/towncrier/_project.py
Comment (adiroiban) [inline src/towncrier/_project.py:104]: I think that we are missing a brach test here

File: src/towncrier/_project.py
Comment (adiroiban) [inline src/towncrier/_project.py:90]: I don't know why coverage is reporting this line as not covered.... we need to check

File: src/towncrier/test/test_project.py
Comment (sigma67) [reply src/towncrier/test/test_project.py]:
```suggestion
        An exception is raised when the version could not be detected.
        For this test we use an incremental-like object,
        that has the `base` method, but that method
        does not match the return type for `incremental`.
```
fixed some typos

File: src/towncrier/_project.py
Comment (sigma67) [reply src/towncrier/_project.py:104]: added

File: src/towncrier/test/test_packaging.py
Comment (sigma67) [reply src/towncrier/test/test_packaging.py:10]: The previous test was checking for the presence of an Incremental Version object (which obviously no longer exists in towncrier)

I've replaced it with a test to assert that a `DeprecationWarning` is present

File: src/towncrier/_project.py
Comment (sigma67) [reply src/towncrier/_project.py:90]: There was an indentation problem with a test, fixed now

**Review by sigma67 — COMMENTED** [review]
thanks, mypy and coverage should be passing now

File: src/towncrier/_project.py
Comment (adiroiban) [reply src/towncrier/_project.py:104]: I still see this error reported 

> line 104 didn't return from function 'get_project_name' because the return on line 105 wasn't executed

image

Comment (sigma67) [reply src/towncrier/_project.py:104]: The miss indicator was a bit confusing for me, should be gone now hopefully

**Comment by sigma67** [comment]
@adiroiban just to know a timeline, is there any ETA on the next release that could include this change?

This is just a question, no pressure

**Comment by adiroiban** [comment]
We can do a release.
There is already a ticket for that here https://github.com/twisted/towncrier/issues/616

the important part is to have someone who cares for a new release and has some time to test the release

**Comment by adiroiban** [comment]
Thanks for fixing all the checks.

As part of this PR we need to update the RELEASE.rst ... if we just merge, the information for RELEASE.rst will no longer be valid

I now see that RELEASE.rst changes were removed from this PR

https://github.com/twisted/towncrier/pull/491/files#diff-0a92214d83ea13ad864f4b1d6c0b5703956cf4bd927c116e75f68d4048e0744f

--------

My hope was that with this PR we can have `pyproject.toml` with

```
[metadata]
version = attr: towncrier.__version__
```

**Comment by adiroiban** [comment]
I made a few small changes to the PR description... the important one is linking to the issue, so that when we merge this PR, the issue is automatically closed.

**Comment by sigma67** [comment]
Ok, I'm a bit confused admittedly because in the prior MR it was stated explicitly to keep the version file.

Here: https://github.com/twisted/towncrier/pull/491#discussion_r1221613227

If we still want to move to pyproject.toml I'd prefer to do that in a separate MR.

What do you want changed in RELEASE.rst ? Should I update the version as well or is that done by the release manager?

**Comment by adiroiban** [comment]
Right now the content of the RELEASE.rst is

https://github.com/twisted/towncrier/blob/4b58be5da5b1fabf38f47d5b777eb9c328046cda/RELEASE.rst?plain=1#L21-L24


----------

I think it should change to

```rst
 Update the version to the release candidate with the first being ``rc1`` (as opposed to 0). 
 In ``src/towncrier/_version.py`` the version is set using PEP440 a compliant string, such as:: 
  
     __version__ = "24.7.0rc1" 
```

---------

This revert is not quite OK https://github.com/twisted/towncrier/pull/627/commits/2cf4e76af7f16d8125118cf4fd47f0747741573f

Instead of updating the pyproject.toml we update the `_version.py` file.

We no longer want to use `incremental` . This is why `towncrier` own development documentation should no longer instruct developers to use `incremental`


---------

But I understand your concern.

The question is where to store the version:

* `pyproject.toml`
* `src/towncrier/_version.py`
* `src/towncier/__init__.py`

-----------

Since the version was already stored in `src/towncrier/_version.py` , my suggestion for this PR is to continue to store it there.

In a separate PR we can consider moving it inside `pyproject.toml` or somewhere else.

**Comment by sigma67** [comment]
Ok my bad, I guess it only needed some changes instead of a revert. RELEASE.rst is updated.

**Review by adiroiban — APPROVED** [review]
Many thanks for the update.

---

I did a few change to release.rst

There was a section underline mismatch ... so I start doing the edit and then I realized that 19.9.0 is a better example as it shows that we don't add leasing 0 in the version ... but we use ISO date format with leading zero.

**Comment by adiroiban** [comment]
I have enabled auto-merge on this.

and once this is merge I plan to do a new release.

Thanks again.

I am very happy to see that we no longer use incremental.

**Pull request merged**
